File: preprocessing/clean_data.py
import pandas as pd


def clean_data(df):
    # Convert character columns to factors (categorical)
    for col in df.select_dtypes(include=['object']).columns:
        df[col] = df[col].astype('category')

    # Replace NA values in numeric columns with the mean of the column
    for col in df.select_dtypes(include=['number']).columns:
        df[col].fillna(df[col].mean(), inplace=True)

    # Filter out rows where the Churn column is NA
    df = df[df['Churn'].notna()]

    return df

File: preprocessing/clean_data.R

library(dplyr)

clean_data <- function(df){

  df <- df %>%
    mutate_if(is.character, as.factor)

  df <- df %>%
    mutate_if(is.numeric, ~replace(., is.na(.), mean(., na.rm = TRUE)))

  df <- df %>%
    filter(!is.na(Churn))

  return(df)
}
